textattack/shared/scripts/run_attack_parallel.py

```diff
@@ -104,28 +109,39 @@ def run(args):
         num_results = 0
         num_failures = 0
         num_successes = 0
-    pbar = tqdm.tqdm(total=num_remaining_examples, smoothing=0)
-    while num_results < num_total_examples:
+    pbar = tqdm.tqdm(total=num_remaining_attacks, smoothing=0)
+    while num_results < num_total_examples and worklist:
         result = out_queue.get(block=True)
 
         if isinstance(result, Exception):
             raise result
+        idx, result = result
         attack_log_manager.log_result(result)
+        worklist.remove(idx)
+
         if (not args.attack_n) or (not isinstance(result, textattack.attack_results.SkippedAttackResult)):
             pbar.update()
             num_results += 1
+
             if type(result) == textattack.attack_results.SuccessfulAttackResult:
                 num_successes += 1
             if type(result) == textattack.attack_results.FailedAttackResult:
                 num_failures += 1
             pbar.set_description('[Succeeded / Failed / Total] {} / {} / {}'.format(num_successes, num_failures, num_results))
         else:
-            label, text = next(dataset)
-            in_queue.put((label, text))
+            last_example += 1
+            try:
+                label, text = dataset[last_example]
+                worklist.append(last_example)
+                in_queue.put((last_example, label, text))
+            except IndexError:
+                last_example -= 1
 
         if args.checkpoint_interval and num_results % args.checkpoint_interval == 0:
             attack_log_manager.flush()
-            checkpoint = textattack.shared.Checkpoint(args, attack_log_manager)
+            checkpoint = textattack.shared.Checkpoint(args, attack_log_manager, worklist, last_example)
+            if not checkpoint.verify_no_duplicates():
```

Comment:
Okay, I have some questions about this.

1. Why are duplicate results allowed? (Why do you warn instead of erroring out?)

2. Why isn't this check completed in the `Checkpoint` constructor?

3. How can this ever happen?

Reply:
I thought maybe it would be better to allow users to have duplicates and have them recover it. But now I think about it, it's highly unlikely and if it happens, i agree we should just error out.